### PR TITLE
Fix Delta Lake insertion issue by setting UTC timezone for timestamp with timezone partition column

### LINKED ISSUE #16822: Delta Lake connector can partition on timestamp with time zone type, but can't insert rows
It would be better to disallow creating such tables or add support. 
```sql
trino:tpch> create table test (c1 int, c2 timestamp with time zone) with (partitioned_by = array['c2']);
CREATE TABLE
trino:tpch> insert into test select 1, current_timestamp;
023-03-31T13:48:25.344+0900	ERROR	remote-task-callback-13	io.trino.execution.scheduler.PipelinedStageExecution	Pipelined stage execution for stage 20230331_044825_00026_573e6.1 failed
io.trino.spi.TrinoException: unsupported type: timestamp(3) with time zone
	at io.trino.plugin.hive.util.HiveWriteUtils.getField(HiveWriteUtils.java:379)
	at io.trino.plugin.hive.util.HiveWriteUtils.createPartitionValues(HiveWriteUtils.java:282)
	at io.trino.plugin.deltalake.AbstractDeltaLakePageSink.createPartitionValues(AbstractDeltaLakePageSink.java:443)
	at io.trino.plugin.deltalake.AbstractDeltaLakePageSink.getWriterIndexes(AbstractDeltaLakePageSink.java:360)
	at io.trino.plugin.deltalake.AbstractDeltaLakePageSink.writePage(AbstractDeltaLakePageSink.java:285)
	at io.trino.plugin.deltalake.AbstractDeltaLakePageSink.appendPage(AbstractDeltaLakePageSink.java:279)
	at io.trino.plugin.base.classloader.ClassLoaderSafeConnectorPageSink.appendPage(ClassLoaderSafeConnectorPageSink.java:69)
	at io.trino.operator.TableWriterOperator.addInput(TableWriterOperator.java:258)
	at io.trino.operator.Driver.processInternal(Driver.java:407)
	at io.trino.operator.Driver.lambda$process$8(Driver.java:305)
	at io.trino.operator.Driver.tryWithLock(Driver.java:701)
	at io.trino.operator.Driver.process(Driver.java:297)
	at io.trino.operator.Driver.processForDuration(Driver.java:268)
	at io.trino.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:845)
	at io.trino.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:165)
	at io.trino.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:537)
	at io.trino.$gen.Trino_testversion____20230331_044759_1.run(Unknown Source)
```

### PR DESCRIPTION
**Pull request by albericgenius**
…with time zone type

## Description
Fix #16822

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Review by ebyhr — COMMENTED** [review]
Please add both normal test and product test.

**Comment by albericgenius** [comment]
> Please add both normal test and product test.

Sure, still working on this.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/AbstractDeltaLakePageSink.java [hunk 1]
```diff
@@ -26,11 +26,11 @@
 import io.trino.parquet.writer.ParquetSchemaConverter;
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/AbstractDeltaLakePageSink.java:26]: Follow the guideline for commit comments for https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#commits-and-pull-requests

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDatabricksInsertCompatibility.java [hunk 2]
```diff
@@ -130,6 +133,40 @@ public void testPartitionedInsertCompatibility()
         }
     }
 
+    @Test(groups = {DELTA_LAKE_DATABRICKS, DELTA_LAKE_OSS, PROFILE_SPECIFIC_TESTS})
+    public void testPartitionedInsertTimestampTZCompatibility()
+    {
+        String tableName = "test_dl_partitioned_insert_timestampTZ" + randomNameSuffix();
+        onTrino().executeQuery("" +
+                "CREATE TABLE delta.default." + tableName +
```
Comment (findinpath) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDatabricksInsertCompatibility.java:141]: Can Spark read the content we inserted with Trino?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/AbstractDeltaLakePageSink.java (hunk 1 above)
Comment (albericgenius) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/AbstractDeltaLakePageSink.java:26]: thanks for information

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDatabricksInsertCompatibility.java (hunk 2 above)
Comment (albericgenius) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDatabricksInsertCompatibility.java:139]: add the spark sql to verify the result, but spark column type is timestamp(not timestamp with time zone)

Comment (ebyhr) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDatabricksInsertCompatibility.java:139]: That's fine. Delta uses timestamp_ntz for timestamp **_without_** time zone type.

**Comment by albericgenius** [comment]
@ebyhr and @findinpath Welcome for any comments

**Comment by albericgenius** [comment]
@ebyhr and @findinpath I guess this is close to getting merged :)

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TransactionLogParser.java [hunk 3]
```diff
@@ -157,7 +180,35 @@ public static Object deserializePartitionValue(DeltaLakeColumnHandle column, Opt
 
     private static Long readPartitionTimestamp(String timestamp)
     {
-        ZonedDateTime zonedDateTime = LocalDateTime.parse(timestamp, PARTITION_TIMESTAMP_FORMATTER).atZone(UTC);
+        ParsePosition position = new ParsePosition(0);
+        TemporalAccessor accessor = PARTITION_TIMESTAMP_FORMATTER.parseUnresolved(timestamp, position);
+        ZonedDateTime zonedDateTime;
+        if (position.getErrorIndex() == -1 && timestamp.length() == position.getIndex()) {
+            if (accessor.isSupported(NANO_OF_SECOND)) {
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TransactionLogParser.java:187]: These conditions look hard to understand for me. Could you leave a code comment with the example?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/util/DeltaLakeWriteUtils.java [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+import static io.trino.spi.type.VarbinaryType.VARBINARY;
+import static java.lang.Float.intBitsToFloat;
+import static java.lang.Math.toIntExact;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Collections.unmodifiableList;
+import static java.util.Collections.unmodifiableMap;
+
+public final class DeltaLakeWriteUtils
+{
+    private DeltaLakeWriteUtils() {}
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/util/DeltaLakeWriteUtils.java:72]: What's the difference from `HiveWriteUtils`? Please leave a code comment.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDatabricksInsertCompatibility.java [hunk 5]
```diff
@@ -130,6 +131,37 @@ public void testPartitionedInsertCompatibility()
...
+                "         (c1 INT, c2 TIMESTAMP WITH TIME ZONE)" +
```
Comment (ebyhr) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDatabricksInsertCompatibility.java:140]: Please add a test for partitioned by nested type with timestamp with time zone.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDatabricksInsertCompatibility.java [hunk 6]
```diff
@@ -130,6 +131,37 @@ public void testPartitionedInsertCompatibility()
...
+                "         WITH (" +
+                "         partitioned_by = ARRAY['c2']," +
+                "         location = 's3://" + bucketName + "/databricks-compatibility-test-" + tableName + "')");
+
+        try {
+            onTrino().executeQuery("INSERT INTO delta.default." + tableName + " VALUES (1, TIMESTAMP '2023-04-05 10:00:00.666+01:00')");
+            onTrino().executeQuery("INSERT INTO delta.default." + tableName + " VALUES (2, TIMESTAMP '2023-04-06 10:00:00+01:00')");
```
Comment (ebyhr) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDatabricksInsertCompatibility.java:147]: nit: No need to split into two INSERT statements.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDatabricksInsertCompatibility.java [hunk 7]
```diff
@@ -130,6 +131,37 @@ public void testPartitionedInsertCompatibility()
...
+            List<Row> expectedRows = ImmutableList.of(
+                    row(
+                            1,
+                            Timestamp.valueOf("2023-04-05 09:00:00.666")),
+                    row(
+                            2,
+                            Timestamp.valueOf("2023-04-06 09:00:00")));
+            assertThat(onTrino().executeQuery("SELECT c1, CAST(c2 AS TIMESTAMP) FROM delta.default." + tableName + " ORDER BY c1 ASC"))
```
Comment (ebyhr) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDatabricksInsertCompatibility.java:155]: Remove redundant `ORDER BY`

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDatabricksInsertCompatibility.java [hunk 8]
```diff
@@ -130,6 +131,37 @@ public void testPartitionedInsertCompatibility()
         }
     }
 
+    @Test(groups = {DELTA_LAKE_DATABRICKS, DELTA_LAKE_OSS, PROFILE_SPECIFIC_TESTS})
+    public void testPartitionedInsertTimestampTZCompatibility()
```
Comment (ebyhr) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDatabricksInsertCompatibility.java:135]:
```suggestion
    public void testPartitionedInsertTimestampWithTimeZoneCompatibility()
```

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/util/DeltaLakeWriteUtils.java [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+                            "Hive partition keys can only contain printable ASCII characters (0x20 - 0x7E). Invalid value: " +
+                                    base16().withSeparator(" ", 2).encode(valueString.getBytes(UTF_8)));
+                }
+                partitionValues.add(valueString);
+            }
+        }
+        return partitionValues.build();
+    }
+
+    public static Object getField(DateTimeZone localZone, Type type, Block block, int position)
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/util/DeltaLakeWriteUtils.java:95]: All types in this method are required in Delta Lake connector?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/util/DeltaLakeWriteUtils.java [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+        long millisUtc = unpackMillisUtc(value);
+        if (type.isShort()) {
+            Instant instant = Instant.ofEpochMilli(millisUtc);
+            return new TimestampTZ((long) Math.floor(millisUtc / 1000), instant.getNano(), UTC_KEY.getZoneId());
+        }
+        else {
+            int picosOfMilli = block.getInt(position, SIZE_OF_LONG);
+            Instant instant = Instant.ofEpochMilli(millisUtc).plusNanos(picosOfMilli * 1000);
+            return new TimestampTZ(instant.getEpochSecond(), instant.getNano(), UTC_KEY.getZoneId());
+        }
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/util/DeltaLakeWriteUtils.java:193]: Is this condition used in Delta Lake connector?

**Review by ebyhr — COMMENTED** [review]
Please squash commits into one.

**Comment by ebyhr** [comment]
@albericgenius Are you still working on this?

**Comment by ebyhr** [comment]
Superseded by #18353

**Pull request closed without merging**